### PR TITLE
RavenDB-12125

### PR DESCRIPTION
**Pull request by ppekrol**
- during database dispose we should prevent handling of any new state changed notifications and wait for a short period of time for current notification to finish processing
- early exiting IndexStore RecordChanged process when shutdown is requested

### CODE REVIEW

**Comment by ppekrol** [comment]
cc @karmeli87

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 1]
```diff
@@ -416,6 +417,11 @@ private unsafe void DisposeInternal()
 
             var exceptionAggregator = new ExceptionAggregator(_logger, $"Could not dispose {nameof(DocumentDatabase)} {Name}");
 
+            var lockTaken = false;
+            Monitor.TryEnter(_locker, TimeSpan.FromSeconds(5), ref lockTaken);
+            if (lockTaken == false && _logger.IsOperationsEnabled)
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentDatabase.cs]: Let's wrap this in a `try/finally`, to be safe.

Comment (ayende) [reply src/Raven.Server/Documents/DocumentDatabase.cs]: I know that we are using the aggregator here, but it is possible that other things could throw that aren't expected and it bugs me that the lock may be leaked.

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/DocumentDatabase.cs (hunk 1 above)
Comment (ppekrol) [reply src/Raven.Server/Documents/DocumentDatabase.cs]: Done

**Pull request merged**